src/network/ethereum.rs

```diff
@@ -1,87 +1,143 @@
+impl Default for PevmChainEthereum {
+    /// Ethereum Mainnet
+    fn default() -> Self {
+        Self {
+            id: NamedChain::Mainnet.into(),
+        }
+    }
 }
 
-pub(crate) fn get_gas_price(tx: &Transaction) -> Result<U256, TransactionParsingError> {
-    let tx_type_raw: u8 = tx.transaction_type.unwrap_or_default();
-    let Ok(tx_type) = TxType::try_from(tx_type_raw) else {
-        return Err(TransactionParsingError::InvalidType(tx_type_raw));
-    };
-
-    match tx_type {
-        TxType::Legacy | TxType::Eip2930 => tx
-            .gas_price
-            .map(U256::from)
-            .ok_or(TransactionParsingError::MissingGasPrice),
-        TxType::Eip1559 | TxType::Eip4844 => tx
-            .max_fee_per_gas
-            .map(U256::from)
-            .ok_or(TransactionParsingError::MissingMaxFeePerGas),
+/// Error type for [PevmEthereum::get_block_spec].
+#[derive(Debug, Clone)]
+pub enum GetBlockSpecError {
+    /// When [header.number] is none.
+    MissingBlockNumber,
+    /// When [header.total_difficulty] is none.
+    MissingTotalDifficulty,
+}
+
+/// Error type for [PevmEthereum::get_gas_price].
+#[derive(Debug, Clone)]
+pub enum GetGasPriceError {
+    /// [tx.type] is invalid.
+    InvalidType(u8),
+    /// [tx.gas_price] is none.
+    MissingGasPrice,
+    /// [tx.max_fee_per_gas] is none.
+    MissingMaxFeePerGas,
+}
+
+impl PevmChain for PevmChainEthereum {
+    type BuildMvMemoryError = Infallible;
+    type GetBlockSpecError = GetBlockSpecError;
+    type GetGasPriceError = GetGasPriceError;
+
+    fn id(&self) -> u64 {
+        self.id
+    }
+
+    fn build_mv_memory(
+        hasher: &ahash::RandomState,
+        block_env: &BlockEnv,
+        txs: &[TxEnv],
+    ) -> Result<MvMemory, Infallible> {
+        let block_size = txs.len();
+        let beneficiary_location_hash = hasher.hash_one(MemoryLocation::Basic(block_env.coinbase));
+
+        // TODO: Estimate more locations based on sender, to, etc.
+        let mut estimated_locations = HashMap::with_hasher(BuildIdentityHasher::default());
+        estimated_locations.insert(
+            beneficiary_location_hash,
+            (0..block_size).collect::<Vec<TxIdx>>(),
+        );
+
+        let mut lazy_addresses = LazyAddresses::default();
+        lazy_addresses.0.insert(block_env.coinbase);
+
+        Ok(MvMemory::new(
+            block_size,
+            estimated_locations,
+            lazy_addresses,
+        ))
+    }
+
+    /// Get the REVM spec id of an Alloy block.
+    // Currently hardcoding Ethereum hardforks from these reference:
+    // https://github.com/paradigmxyz/reth/blob/4fa627736681289ba899b38f1c7a97d9fcf33dc6/crates/primitives/src/revm/config.rs#L33-L78
+    // https://github.com/paradigmxyz/reth/blob/4fa627736681289ba899b38f1c7a97d9fcf33dc6/crates/primitives/src/chain/spec.rs#L44-L68
+    // TODO: Better error handling & properly test this.
+    fn get_block_spec(header: &Header) -> Result<SpecId, Self::GetBlockSpecError> {
```

Comment:
Should add a comment to `PevmEthereum` that we only support mainnet at the moment.